akka-projection-grpc/src/main/scala/akka/projection/grpc/replication/internal/ReplicationImpl.scala

```diff
@@ -0,0 +1,215 @@
+import akka.cluster.sharding.typed.ShardedDaemonProcessSettings
+import akka.cluster.sharding.typed.scaladsl.ClusterSharding
+import akka.cluster.sharding.typed.scaladsl.Entity
+import akka.cluster.sharding.typed.scaladsl.EntityRef
+import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+import akka.cluster.sharding.typed.scaladsl.ShardedDaemonProcess
+import akka.http.scaladsl.model.HttpRequest
+import akka.http.scaladsl.model.HttpResponse
+import akka.persistence.Persistence
+import akka.persistence.query.typed.EventEnvelope
+import akka.persistence.typed.PublishedEvent
+import akka.persistence.typed.ReplicationId
+import akka.persistence.typed.internal.PublishedEventImpl
+import akka.persistence.typed.internal.ReplicatedEventMetadata
+import akka.persistence.typed.internal.ReplicatedPublishedEventMetaData
+import akka.persistence.typed.scaladsl.EventSourcedBehavior
+import akka.persistence.typed.scaladsl.ReplicatedEventSourcing
+import akka.persistence.typed.scaladsl.ReplicationContext
+import akka.projection.ProjectionBehavior
+import akka.projection.ProjectionContext
+import akka.projection.ProjectionId
+import akka.projection.eventsourced.scaladsl.EventSourcedProvider
+import akka.projection.grpc.consumer.GrpcQuerySettings
+import akka.projection.grpc.consumer.scaladsl.GrpcReadJournal
+import akka.projection.grpc.producer.scaladsl.EventProducer
+import akka.projection.grpc.producer.scaladsl.EventProducer.EventProducerSource
+import akka.projection.grpc.producer.scaladsl.EventProducer.Transformation
+import akka.projection.grpc.replication.scaladsl.Replica
+import akka.projection.grpc.replication.scaladsl.Replication
+import akka.stream.scaladsl.FlowWithContext
+import akka.util.Timeout
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class ReplicationImpl[Command] private (
+    val eventProducerService: EventProducerSource,
+    val createSingleServiceHandler: () => PartialFunction[HttpRequest, Future[HttpResponse]],
+    val entityTypeKey: EntityTypeKey[Command],
+    val entityRefFactory: String => EntityRef[Command])
+    extends Replication[Command] {
+  override def toString: String = s"Replication(${eventProducerService.entityType}, ${eventProducerService.streamId})"
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object ReplicationImpl {
+
+  private val log = LoggerFactory.getLogger(classOf[ReplicationImpl[_]])
+
+  private val filteredEvent = Future.successful(None)
+
+  /**
+   * Called to bootstrap the entity on each cluster node in each of the replicas.
+   *
+   * Important: Note that this does not publish the endpoint, additional steps are needed!
+   */
+  def grpcReplication[Command, Event, State](settings: ReplicationSettingsImpl[Command])(
+      replicatedBehaviorFactory: ReplicationContext => EventSourcedBehavior[Command, Event, State])(
```

Comment:
One problem with this type of factory is that since it needs to return an event sourced behavior it can't be composed to access context and timers

Reply:
Peter has a solution for that in #769